### PR TITLE
[LLVM][SLSR] Add a debug counter 

### LINKED ISSUE #119770: [LLVM][SLSR] Add a debug counter
Add a debug counter to selectively enable or disable SLSR transformations. The debug counter should wrap `SLSR`'s invocation of [rewriteCandidateWithBasis()](https://github.com/llvm/llvm-project/blob/bd40421ad9ec5ecc164f8208caf3ba5657977e17/llvm/lib/Transforms/Scalar/StraightLineStrengthReduce.cpp#L693C7-L693C32).

---

See [DAGCombiner.cpp](https://github.com/llvm/llvm-project/blob/bd40421ad9ec5ecc164f8208caf3ba5657977e17/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp#L2025-L2026) for an example use case for a debug counter:
```
SDValue DAGCombiner::combine(SDNode *N) {
  if (!DebugCounter::shouldExecute(DAGCombineCounter))
      return SDValue();
  ...
```

See 7a8dc3dafa85a1bbbb9458f2162ac5d61a9dc91a for an example of a patch that adds support for a debug counter. 

---

Don't feel the need to ask permission to work on this. Just give it a try! 

CC @akshayrdeodhar

### PR DESCRIPTION
**Pull request by GrumpyPigSkin**
Added debug counter and test for slsr pass

fixes: https://github.com/llvm/llvm-project/issues/119770

### CODE REVIEW

**Comment by GrumpyPigSkin** [comment]
@justinfargnoli I can't add you to the reviewers for some reason so giving you a mention here. Let me know if I need to make any changes.

File: llvm/test/Other/debugcounter-slsr.ll [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  %s4 = shl i32 %s, 2
+  %t2 = add i32 %b, %s4
+  call void @foo(i32 %t2)
+  %s6 = mul i32 %s, 6
+  %t3 = add i32 %b, %s6
+  call void @foo(i32 %t3)
+  ret void
+}
+
+declare void @foo(i32)
```
Comment (justinfargnoli) [inline llvm/test/Other/debugcounter-slsr.ll:30]: Add a newline to the end of the file.

File: llvm/test/Other/debugcounter-slsr.ll [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+;
+  %s2 = shl i32 %s, 1
+  %t1 = add i32 %b, %s2
+  call void @foo(i32 %t1)
+  %s4 = shl i32 %s, 2
+  %t2 = add i32 %b, %s4
+  call void @foo(i32 %t2)
+  %s6 = mul i32 %s, 6
+  %t3 = add i32 %b, %s6
+  call void @foo(i32 %t3)
```
Comment (justinfargnoli) [inline llvm/test/Other/debugcounter-slsr.ll:26]: To reduce the size of the test case, lets delete this code.

File: llvm/test/Other/debugcounter-slsr.ll [hunk 3]
```diff
@@ -0,0 +1,30 @@
+; REQUIRES: asserts
+; RUN: opt -passes=slsr,gvn -S -debug-counter=slsr-counter=1  < %s | FileCheck %s
+
+;; Test that, with debug counters on, we will skip the first slsr opportunity.
```
Comment (justinfargnoli) [inline llvm/test/Other/debugcounter-slsr.ll:4]: `;;` -> `;`

**Review by justinfargnoli — COMMENTED** [review]
Overall, this looks great! 

Leaving a few small "touch-up" comments.

**Comment by GrumpyPigSkin** [comment]
@justinfargnoli, I have resolved all the comments. I had more of an issue adding a newline to the end of the file than anything 🙄 It Looks like the pull request has picked up the changes, let me know if there are any issues :)

File: llvm/test/Other/debugcounter-slsr.ll [hunk 4]
```diff
@@ -0,0 +1,26 @@
+; REQUIRES: asserts
+; RUN: opt -passes=slsr,gvn -S -debug-counter=slsr-counter=1  < %s | FileCheck %s
```
Comment (justinfargnoli) [inline llvm/test/Other/debugcounter-slsr.ll:2]: Do we need to run GVN (Global Value Numbering) here? It'd be nicer if you could eliminate the need to run it.

Comment (GrumpyPigSkin) [reply llvm/test/Other/debugcounter-slsr.ll]: No, we don't I have removed it.

**Review by justinfargnoli — APPROVED** [review]
Great work on this @GrumpyPigSkin! LGTM :)

File: llvm/lib/Transforms/Scalar/StraightLineStrengthReduce.cpp [hunk 5]
```diff
@@ -690,7 +694,8 @@ bool StraightLineStrengthReduce::runOnFunction(Function &F) {
   while (!Candidates.empty()) {
     const Candidate &C = Candidates.back();
     if (C.Basis != nullptr) {
-      rewriteCandidateWithBasis(C, *C.Basis);
+      if (DebugCounter::shouldExecute(StraightLineStrengthReduceCounter))
+        rewriteCandidateWithBasis(C, *C.Basis);
```
Comment (justinfargnoli) [inline llvm/lib/Transforms/Scalar/StraightLineStrengthReduce.cpp:698]: Apologies for the last-minute review. 

After looking through other uses of debug counters, it seems like `shouldExecute` is usually called within the function that implements the transformation as opposed to guarding the call to the function that implements the transformation. 

i.e. 
```
if (shouldExecute())
  foo();
-->
void foo() {
  if (shouldExecute())
    return;
  ...
```

Can we make this change and then go forward with merging the PR?

Comment (GrumpyPigSkin) [reply llvm/lib/Transforms/Scalar/StraightLineStrengthReduce.cpp]: Good point, I have moved it to the top of `rewriteCandidateWithBasis` :)

File: llvm/lib/Transforms/Scalar/StraightLineStrengthReduce.cpp [hunk 6]
```diff
@@ -267,9 +271,10 @@ FunctionPass *llvm::createStraightLineStrengthReducePass() {
 
 bool StraightLineStrengthReduce::isBasisFor(const Candidate &Basis,
                                             const Candidate &C) {
-  return (Basis.Ins != C.Ins && // skip the same instruction
-          // They must have the same type too. Basis.Base == C.Base doesn't
-          // guarantee their types are the same (PR23975).
+  return (Basis.Ins !=
+              C.Ins && // skip the same instruction
+                       // They must have the same type too. Basis.Base == C.Base
+                       // doesn't guarantee their types are the same (PR23975).
```
Comment (justinfargnoli) [inline llvm/lib/Transforms/Scalar/StraightLineStrengthReduce.cpp:277]: Please delete this diff, then you're good to go!

Comment (GrumpyPigSkin) [reply llvm/lib/Transforms/Scalar/StraightLineStrengthReduce.cpp]: Sorted it sorry, happened when I used the `git clang-format HEAD~1` command

Comment (justinfargnoli) [reply llvm/lib/Transforms/Scalar/StraightLineStrengthReduce.cpp]: Ahh, interesting. I didn't realize that's where the diff came from.

**Comment by justinfargnoli** [comment]
Merging on @GrumpyPigSkin's behalf because he doesn't have commit access.

**Comment by justinfargnoli** [comment]
> LLVM Buildbot has detected a new failure on builder publish-sphinx-docs running on as-worker-4 while building llvm at step 7 "Publish docs-clang-html".

This looks like an infra failure and is likely unrelated to this PR.

**Pull request merged**